src/or/onion.c

```diff
@@ -1339,3 +1342,466 @@ extended_cell_format(uint8_t *command_out, uint16_t *len_out,
+ */
+create2v_cell_t*
+create2v_cell_new(void)
+{
+  create2v_cell_t *cell;
+
+  cell = tor_malloc(sizeof(create2v_cell_t));
+  cell->body = create2v_cell_body_new();
+
+  return cell;
+}
+
+/**
+ * Conduct various length checks on some handshake and padding data lengths for
+ * a given <b>handshake_type</b>.
+ *
+ * If <b>is_initiator</b> is true, then lengths are checked for the Alice side
+ * of the handshake, otherwise, if false, the checks are run for Bob's
+ * response.
+ */
+// XXXisis should this instead return bool and report errors rather than
+// asserting?
+static void
+check_handshake_and_padding_len(const uint16_t handshake_type,
+                                const uint16_t handshake_len,
+                                const uint16_t padding_len,
+                                const bool is_initiator)
+{
+  (void)is_initiator; /* TODO: Check lengths once we have a handshake.--isis */
+
+  switch (handshake_type) {
+    /* TODO: When we have handshake types which use wide cells, we should check
+     * that their lengths match the expected lengths here. --isis */
+  default:
+    /* If we've been given handshake_data, make sure we were given a
+     * length in (0,10240]. */
+    tor_assert(handshake_len > 0);
+    tor_assert(handshake_len <= 10240);
+
+    /* If we've been given both handshake_data and padding_data, make
+     * sure their combined reported lengths are not greater than the
+     * maximum. */
+    tor_assert(handshake_len + padding_len <= 10240);
+  }
+}
+
+/**
+ * Initialise a CREATE2V cell.
+ *
+ * Copies <b>handshake_len</b> bytes of <b>handshake_data</b> into the
+ * <b>cell_out</b>'s body.hdata member, and then copies <b>padding_len</b>
+ * bytes of <b>padding_data</b> into the <b>cell_out</b>'s body.ignored member.
+ * Also sets cell_out->body.htype to <b>handshake_type</b>.
+ */
+void
+create2v_cell_init(create2v_cell_t *cell_out,
+                   const uint16_t handshake_type,
+                   const uint8_t *handshake_data,
+                   const uint16_t handshake_len,
+                   const uint8_t *padding_data,
+                   const uint16_t padding_len)
+{
+  tor_assert(cell_out);
+
+  check_handshake_and_padding_len(handshake_type, handshake_len,
+                                  padding_len, true);
+
+  create2v_cell_body_set_htype(cell_out->body, handshake_type);
+  create2v_cell_body_set_hlen(cell_out->body, handshake_len);
+  create2v_cell_body_setlen_hdata(cell_out->body, (size_t)handshake_len);
+  create2v_cell_body_setlen_ignored(cell_out->body, (size_t)padding_len);
+
+  if (handshake_len)
+    memcpy(create2v_cell_body_getarray_hdata(cell_out->body),
+           handshake_data, handshake_len);
+  if (padding_len)
+    memcpy(create2v_cell_body_getarray_ignored(cell_out->body),
+           padding_data, padding_len);
+}
+
+/**
+ * Deallocate a CREATE2V cell.
+ */
+void
+create2v_cell_free(create2v_cell_t *cell)
```

Comment:
Please use the same conventions as elsewhere in the code: name this function ending with an underscore, and #define create2v_cell_free() as a macro using FREE_AND_NULL.

Also, make sure this function accepts NULL as an input.